x/ecocredit/server/core/create_batch.go

```diff
@@ -73,51 +73,90 @@ func (k Keeper) CreateBatch(ctx context.Context, req *core.MsgCreateBatch) (*cor
 	maxDecimalPlaces := creditType.Precision
 
 	tradableSupply, retiredSupply := math.NewDecFromInt64(0), math.NewDecFromInt64(0)
+
+	// set module address string once for better performance
 	moduleAddrString := k.moduleAddress.String()
+
 	for _, issuance := range req.Issuance {
+		// get and set decimal values of tradable amount and retired amount
 		decs, err := utils.GetNonNegativeFixedDecs(maxDecimalPlaces, issuance.TradableAmount, issuance.RetiredAmount)
 		if err != nil {
 			return nil, err
 		}
-		tradable, retired := decs[0], decs[1]
-
-		tradableString := tradable.String()
-		retiredString := retired.String()
+		tradableAmount, retiredAmount := decs[0], decs[1]
 
 		recipient, _ := sdk.AccAddressFromBech32(issuance.Recipient)
-		if !tradable.IsZero() {
-			tradableSupply, err = tradableSupply.Add(tradable)
+
+		// get the current batch balance of the recipient account
+		// Note: Get because batch balance may or may not already exist
+		// depending on the length of issuance and if recipient is the same
+		balance, err := k.stateStore.BatchBalanceTable().Get(ctx, recipient, batchKey)
+		if err != nil {
+			if ormerrors.IsNotFound(err) {
+				balance = &api.BatchBalance{
+					TradableAmount: "0",
+					RetiredAmount:  "0",
+					EscrowedAmount: "0",
+				}
+			} else {
+				return nil, err
+			}
+		}
```

Comment:
I think I prefer without the utility function so that we only set the fields we need (i.e. each amount) and we aren't passing a table instance to a separate function (I think this is not memory efficient).